src/mjolnir/shortcutbuilder.cc

```diff
@@ -81,21 +81,24 @@ bool EdgesMatch(const graph_tile_ptr& tile, const DirectedEdge* edge1, const Dir
   // TODO - should allow near matches?
   std::vector<std::string> edge1names = tile->GetNames(edge1);
   std::vector<std::string> edge2names = tile->GetNames(edge2);
-  if (edge1names.size() != edge2names.size()) {
+  std::sort(edge1names.begin(), edge1names.end());
+  std::sort(edge2names.begin(), edge2names.end());
+  if (edge1names != edge2names)
     return false;
-  }
-  for (const auto& name1 : edge1names) {
-    bool found = false;
-    for (const auto& name2 : edge2names) {
-      if (name1 == name2) {
-        found = true;
-        break;
-      }
-    }
-    if (!found) {
+
+  // if they have access restrictions those must match (for modes that use shortcuts)
+  if (edge1->access_restriction()) {
+    auto res1 = tile->GetAccessRestrictions(edge1 - tile->directededge(0), kVehicularAccess);
+    auto res2 = tile->GetAccessRestrictions(edge2 - tile->directededge(0), kVehicularAccess);
+    if (res1.size() != res2.size())
       return false;
+    for (size_t i = 0; i < res1.size(); ++i) {
+      if (res1[i].type() != res2[i].type() || res1[i].modes() != res2[i].modes() ||
+          res1[i].value() != res2[i].value())
```

Comment:
i was going to put a comment on this too, i saw that `operator<` didnt sort on `type` that means that it could be that they are the same set of restrictions but one has hazmat first and the other has weight first but would otherwise be the same. i think the best way to fix this is in the operator, by adding type to the consideration. i started doing it in another tab but didnt commit it yet..

Reply:
done

Reply:
I didn't find code that uses the operator though.. Probably I'm missing some implicit sorting somewhere? I guess at this point I'll need to add some more tests

Reply:
i think all we need is to run the same graph but without the restrictions and assert that we only get one pair of shortcuts for each set of connected ways

Reply:
this is where that operator is used: https://github.com/valhalla/valhalla/blob/ffe790ddfe08b89c02c419ec08cfec0bf715a64c/src/mjolnir/graphtilebuilder.cc#L287

Reply:
Great, thanks, missed that!

Reply:
std::sort is templatized so the compiler looks for `operator<` on whatever thing is used in the template (techincally it looks first for the `operator*` to dereference the iterable thing and then it looks for the less than). in this case the template parameter is implicit as well :upside_down_face: